sparrow_datums/keypoints/keypoints.py

```diff
@@ -0,0 +1,191 @@
+            raise ValidationError(f"Unknown box parameterization: {self.ptype.name}")
+
+    @property
+    def x(self) -> FloatArray:
+        """Slice the x dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 0]
+        return result
+
+    @property
+    def y(self) -> FloatArray:
+        """Slice the y dimension of the keypoint."""
+        self.validate_known_ptype()
+        result: FloatArray
+        result = self.array[..., 1]
+        return result
+
+    def generate_heatmap(self, x0: int, y0: int, covariance: float):
+        """Create a 2D heatmap from an x, y pixel location.
+
+        Note: This is a helper function for to_heatmap_array()
+        Parameters
+        ----------
+        x0 :
+            x coordinate of the keypoint to be transformed
+        y0 :
+            y coordinate of the keypoint to be transformed
+        covariance :
+            covariance of the surface to be created
+        Returns
+        -------
+            keypoint in form of a 2D array( a heatmap)
+        """
+        img_w = self.image_width
+        img_h = self.image_height
+        xx, yy = np.meshgrid(np.arange(img_w), np.arange(img_h))
+        zz = (
+            1
+            / (2 * np.pi * covariance**2)
+            * np.exp(
+                -(
+                    (xx - x0) ** 2 / (2 * covariance**2)
+                    + (yy - y0) ** 2 / (2 * covariance**2)
+                )
+            )
+        )
+        # Normalize zz to be in [0, 1]
+        zz_min = zz.min()
+        zz_max = zz.max()
+        zz_range = zz_max - zz_min
+        if zz_range == 0:
+            zz_range += 1e-8
+        return (zz - zz_min) / zz_range
+
+    def to_heatmap_array(self, covariance: float = 20) -> npt.NDArray[np.float64]:
+        """Convert Keypoints into a heatmap array.
+
+        Parameters
+        ----------
+        covariance :
+            Covariance of the surface to be created, by default 20
+        Returns
+        -------
+            heatmaps in numpy form.
+        """
+        xs = self.array[..., 0]
+        ys = self.array[..., 1]
+        if np.size(xs) > 1:
+            heatmaps = []
+            for x0, y0 in zip(xs, ys):
+                heatmap = self.generate_heatmap(x0, y0, covariance)
+                heatmaps.append(heatmap)
+            return np.stack(heatmaps)
+        elif np.size(xs) == 1:
+            return self.generate_heatmap(xs.item(), ys.item(), covariance)
+
+    @classmethod
+    def from_heatmap_array(
+        self, heatmaps: npt.NDArray[np.float64]
+    ) -> npt.NDArray[np.float64]:
+        """Convert a heatmap array back into Keypoint form.
+
+        Parameters
+        ----------
+        heatmaps :
+            Keypoints represented in form of a 2D surface.
+
+        Returns
+        -------
+            Keypoints in xy coordinates
+        """
+        keypoints = []
```

Comment:
Same issue here -- this should either work with multiple dimensions of heatmaps or we should explicitly throw an error saying we only support the 2D case.

Reply:
Again, I couldn't think of where this could exceed 2D. Especially, heatmap is derived from a keypoint; not independently defined. I just added and tested an exception just to be sure. Let me know if you have any ideas for specific cases.